File: crawler/spiders/stock_futures_list_spider.py
# encoding: utf-8
# pylint: disable=E1101
# 股票期貨清單
# https://www.taifex.com.tw/cht/2/stockLists


import scrapy
from scrapy import signals, Spider
from urllib.parse import urlencode
import time
from random import randint
import datetime
import logging
from copy import copy
from dateutil.relativedelta import relativedelta
import collections
import json
from bs4 import BeautifulSoup
from random import randint
from zoneinfo import ZoneInfo

# for cloud function call && scrapy crawl command call
# softlink package folder to root
try:
    from package.tools import is_settle, format_number, getDateObj
    from package.storage import Storage
except:
    from spiders.package.tools import is_settle, format_number, getDateObj
    from spiders.package.storage import Storage



class StockFuturesListSpider(scrapy.Spider):
    name = 'stock_futures_list'
        
    def __init__(self, category=None, *args, **kwargs):
        super(StockFuturesListSpider, self).__init__(*args, **kwargs)
        
        self.dataStorage = Storage(self.name)
        self.data = collections.OrderedDict()
        self.url = 'https://www.taifex.com.tw/cht/2/stockLists'


    def start_requests(self):
        print('start request - %s' % self.name)
        time.sleep(randint(2,3))
        yield scrapy.FormRequest(
            url=self.url,
            callback=self.parse,
            errback=self.handle_failure)

    def handle_failure(self, failure):
        self.log(failure, level=logging.ERROR)
        # try with a new proxy
        self.log('restart from the failed url {}'.format(failure.request.url))
        time.sleep(120)
        yield scrapy.FormRequest(
            url=failure.request.url,
            callback=self.parse,
            errback=self.handle_failure)

    def parse(self, response):
        soup = BeautifulSoup(response.text, "lxml")
        table = soup.select_one("#myTable tbody")
        
        self.stockList = []
        try:
            trList = table.select("tr")

            for tr in trList:
                stockData = {}
 
                stockData['productCode'] = tr.select("td")[0].text.strip()
                stockData['code'] = tr.select("td")[2].text.strip()
                stockData['name'] = tr.select("td")[3].text.strip()
                stockData['isFutures'] = True if len(tr.select("td")[4].contents) > 1 else False
                stockData['isOption'] = True if len(tr.select("td")[5].contents) > 1 else False
                stockData['stockNumber'] = tr.select("td")[9].text.strip()

                self.stockList.append(stockData)

        except Exception as e:
            print('error - %s' % self.name)
            print(e)


    @classmethod
    def from_crawler(cls, crawler, *args, **kwargs):
        spider = super(StockFuturesListSpider, cls).from_crawler(crawler, *args, **kwargs)
        crawler.signals.connect(spider.spider_closed, signal=signals.spider_closed)
        return spider

    def spider_closed(self, spider):
        fileName = self.name
        self.dataStorage.saveData(fileName, self.stockList)

File: node-crawler/src/crawlers/stock-futures-list-spider.ts
// 股票期貨清單
// https://www.taifex.com.tw/cht/2/stockLists

import * as cheerio from 'cheerio'
import Api from '../api/index.ts'
import Storage from '../storage/index.ts'

type StockData = {
  productCode: string
  code: string
  name: string
  isFutures: boolean
  isOption: boolean
  stockNumber: string
}

const name = 'stock_futures_list'
const dataStorage = new Storage(name)
const stockList: StockData[] = []

function parse(htmlString: string) {
  const $ = cheerio.load(htmlString)
  const table = $('#myTable tbody')

  try {
    const trList = table.find('tr')

    trList.each((i, tr) => {
      const stockData: StockData = {
        productCode: $(tr).find('td').eq(0).text().trim(),
        code: $(tr).find('td').eq(2).text().trim(),
        name: $(tr).find('td').eq(3).text().trim(),
        isFutures: $(tr).find('td').eq(4).contents().length > 1,
        isOption: $(tr).find('td').eq(5).contents().length > 1,
        stockNumber: $(tr).find('td').eq(10).text().trim()
      }

      stockList.push(stockData)
    })
  } catch (e) {
    console.log('error - %s', name)
    console.log(e)
  }
}

async function save() {
  await dataStorage.saveData<StockData[]>(name, stockList)
}

async function main() {
  console.log(`start - ${name}`)

  try {
    const { data: htmlText } = await Api.getStockFuturesList()
    parse(htmlText)
    await save()
  } catch (error) {
    console.error(error)
  }
}

await main()
